src/main/scala/is/hail/io/bgen/BgenRDDPartitions.scala

```diff
@@ -208,5 +236,312 @@ object BgenRDDPartitions extends Logging {
+        cbfis.invoke[Int, Unit]("readLengthAndSkipString", 2)
+      },
+      contig := cbfis.invoke[Int, String]("readLengthAndString", 2),
+      contigRecoded := csettings.invoke[String, String]("recodeContig", contig),
+      position := cbfis.invoke[Int]("readInt"),
+      if (settings.skipInvalidLoci) {
+        Code(
+          invalidLocus :=
+            (if (settings.rg.nonEmpty) {
+              !csettings.invoke[Option[ReferenceGenome]]("rg")
+                .invoke[ReferenceGenome]("get")
+                .invoke[String, Int, Boolean]("isValidLocus", contigRecoded, position)
+            } else false),
+          invalidLocus.mux(
+            Code(
+              nAlleles := cbfis.invoke[Int]("readShort"),
+              i := 0,
+              Code.whileLoop(i < nAlleles,
+                cbfis.invoke[Int, Unit]("readLengthAndSkipString", 4),
+                i := i + 1
+              ),
+              dataSize := cbfis.invoke[Int]("readInt"),
+              Code.toUnit(cbfis.invoke[Long, Long]("skipBytes", dataSize.toL)),
+              Code._return(-1L) // return -1 to indicate we are skipping this variant
+            ),
+            Code._empty // if locus is valid, continue
+          ))
+      } else {
+        if (settings.rg.nonEmpty) {
+          // verify the locus is valid before continuing
+          csettings.invoke[Option[ReferenceGenome]]("rg")
+            .invoke[ReferenceGenome]("get")
+            .invoke[String, Int, Unit]("checkLocus", contigRecoded, position)
+        } else {
+          Code._empty // if locus is valid continue
+        }
+      },
+      srvb.start(),
+      srvb.addBaseStruct(settings.typ.types(settings.typ.fieldIdx("locus")).fundamentalType.asInstanceOf[TBaseStruct], { srvb =>
+        Code(
+          srvb.start(),
+          srvb.addString(contigRecoded),
+          srvb.advance(),
+          srvb.addInt(position))
+      }),
+      srvb.advance(),
+      nAlleles := cbfis.invoke[Int]("readShort"),
+      nAlleles.cne(2).mux(
+        Code._fatal(
+          const("Only biallelic variants supported, found variant with ")
+            .concat(nAlleles.toS)),
+        Code._empty),
+      srvb.addArray(settings.typ.types(settings.typ.fieldIdx("alleles")).asInstanceOf[TArray], { srvb =>
+        Code(
+          srvb.start(nAlleles),
+          i := 0,
+          Code.whileLoop(i < nAlleles,
+            srvb.addString(cbfis.invoke[Int, String]("readLengthAndString", 4)),
+            srvb.advance(),
+            i := i + 1))
+      }),
+      srvb.advance(),
+      if (settings.rowFields.rsid)
+        Code(srvb.addString(rsid), srvb.advance())
+      else Code._empty,
+      if (settings.rowFields.varid)
+        Code(srvb.addString(varid), srvb.advance())
+      else Code._empty,
+      if (settings.rowFields.fileRowIndex)
+        Code(srvb.addLong(fileRowIndex), srvb.advance())
+      else Code._empty,
+      dataSize := cbfis.invoke[Int]("readInt"),
+      settings.entries match {
```

Comment:
I think the stuff before this shouldn't be staged, but I'll let the you and the reviewer make the final call.

Reply:
I don't care much. It wasn't much work to transform it to this, it's not much work to transform it back. I'm mildly inclined to do nothing since that's the path of least resistance.